# Goal: Identify N-DEGs using 'leave-out-one' approach: leaving out one genotype each round and repeat for 18 genotypes 
# Input: read counts generated by featureCounts
# Output: the N-DEG intersected by 18 lists
# Conclusion
# Author: Chia-Yi Cheng
# Last updated: 2021-01-09

# 0. Environment ----------------------------------------------------------------------------------------

library(RUVSeq)
library(edgeR)

rm(list = setdiff(ls(), lsf.str()))
setwd("")

# 1. Load Data ------------------------------------------------------------------------------------------

load(file = "./NutriNet.AGPv4_Leaf.RUVseq-edgeR-202011-24054background-input.RData")

## Filter data by CPM using edgeR

raw <- DGEList(counts=leaf,group=rep,remove.zeros = T)
nrow(raw) #39183
keep <- rowSums(cpm(raw)>1) >= 10 # at least 1 CPM in 10 or more samples 
filtered <- raw[keep, keep.lib.sizes=T] # 24054 genes left.
nrow(filtered)
universe <- rownames(filtered)

## Normalization
set <- newSeqExpressionSet(as.matrix(filtered), 
                           phenoData = data.frame(rep, row.names =colnames(filtered)))
set1 <- betweenLaneNormalization(set, which="upper")

## RUVs: Estimating the factors of unwanted variation using replicate samples
differences<-makeGroups(rep)
set3 <- RUVs(set1, universe, k=1, differences)

# 2. DE analysis ------------------------------------------------------------------------------------------

library(edgeR)
library(tidyverse)

# edgeR for loop
n=18 # number of genotype
deg.out = vector('list',n)

# prepare leave out index
leaveoutvectorlist=list(1:6,7:10,11:15,16:20,21:24,25:29,30:34,35:39,40:44,
                       46:49,50:55,56:59,60:61,62:67,68:72,73:78,79:84,85:89)

for (i in 1:n){
  leaveout=unlist(leaveoutvectorlist[i])
  
  Ncondition=sampleinfo$Nitrogen[-leaveout]
  accession=factor(sampleinfo$Genotype[-leaveout])
  rep=factor(sampleinfo$Replicate[-leaveout])
  pdata=pData(set3)[-leaveout,]
  counts=counts(set3)[,-leaveout]
  
  design <- model.matrix(~Ncondition+accession+W_1,data=pdata)
  design
  colnames(design)
  
  y <- DGEList(counts=counts,group=rep)
  y <- calcNormFactors(y, method = 'upperquartile')
  y$samples
  
  y<-estimateDisp(y,design,robust = T)
  y$common.dispersion 
  
  fit <-glmFit(y, design)
  
  # DE by Nitrogen
  lrt <-glmLRT(fit, coef=2)
  top <- topTags(lrt, n=nrow(set3))$table
  
  de.gene <- rownames(top[1:sum(top$FDR<0.05),])
  Pval <- top$PValue[1:sum(top$FDR<0.05)]
  FDR <- top$FDR[1:sum(top$FDR<0.05)]
  deg.out[[i]] <- paste(de.gene,Pval,FDR,sep = ",")

}

save(deg.out,diff.deg,diff.xgboost,annotation,
     file = "./NutriNet.AGPv4_Leaf.RUVseq-edgeR-202011-24054background-output.RData")

# 3. Analysis: Find the DEGs shared by all 18 left-out-one lists ---------------------------------------

library(data.table)
library(tidyverse)

rm(list = setdiff(ls(), lsf.str()))
setwd("")
load(file = "NutriNet.AGPv4_Leaf.RUVseq-edgeR-202011-24054background-output.RData")

## consolidate the 18 lists
degout.df <- as.data.frame(do.call(rbind,flatten(deg.out))) %>% 
  separate(.,V1, into =c("DEG","Pval","FDR"), sep=",")

## count the frequency
deg.df=data.frame(table(degout.df$DEG))
names(deg.df)=c("Gene","Freq")
setDT(deg.df)
deg.df[order(-Freq)]

union<-deg.df$Freq==n
table(union)
DEG.list=factor(deg.df$Gene[union]) #6914

## Calculate mean Pval
degout.keep = degout.df %>% filter(DEG %in% DEG.list)
degout.pval = degout.keep[,c(1,2)]

pval = as.numeric(degout.pval$Pval)
DEG.pval =aggregate(x=pval,by=list(degout.pval$DEG),FUN=mean)

names(DEG.pval) = c("Gene","Pval")
setDT(DEG.pval)
DEG.pval

## Add annotation
DEG.pval.annot=merge(DEG.pval,annotation,by="Gene",all.x = T)[order(Pval)]
DEG.pval.annot

## Add the frequency of occurrance
unionDEG.Pval = data.table(DEG.pval.annot,edgeR.ranking = rownames(DEG.pval))
unionDEG.Pval

save(DEG.list, degout.df,n, unionDEG.Pval,
     file = "./NutriNet.AGPv4_Leaf.RUVseq-edgeR-202011-24054background-analysis.RData")

write.table(DEG.list, quote=FALSE, row.names=F, col.names=F,sep = "\t",
            file="./NutriNet-Zm-leaf.edgeR-N-6914geneID.txt")

write.table(DEG.pval.annot, quote=FALSE, row.names=F, col.names=F,sep = "\t",
            file="./NutriNet-Zm-leaf.edgeR-N.6914geneID-meanPval-annotation.txt")

# 4. Session Information -------------------------------------------------------------------------------
sessionInfo()


#> sessionInfo()
#R version 4.0.3 (2020-10-10)
#Platform: x86_64-w64-mingw32/x64 (64-bit)
#Running under: Windows >= 8 x64 (build 9200)
#
#Matrix products: default
#
#locale:
#[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
#[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
#[5] LC_TIME=English_United States.1252    
#
#attached base packages:
#[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     
#
#other attached packages:
# [1] fdrtool_1.2.15              reshape2_1.4.4              RColorBrewer_1.1-2         
# [4] RUVSeq_1.24.0               EDASeq_2.24.0               ShortRead_1.48.0           
# [7] GenomicAlignments_1.26.0    SummarizedExperiment_1.20.0 MatrixGenerics_1.2.0       
#[10] matrixStats_0.57.0          Rsamtools_2.6.0             GenomicRanges_1.42.0       
#[13] GenomeInfoDb_1.26.0         Biostrings_2.58.0           XVector_0.30.0             
#[16] IRanges_2.24.0              S4Vectors_0.28.0            BiocParallel_1.24.0        
#[19] Biobase_2.50.0              BiocGenerics_0.36.0         WGCNA_1.69                 
#[22] fastcluster_1.1.25          dynamicTreeCut_1.63-1       edgeR_3.32.0               
#[25] limma_3.46.0                forcats_0.5.0               stringr_1.4.0              
#[28] dplyr_1.0.2                 purrr_0.3.4                 readr_1.4.0                
#[31] tidyr_1.1.2                 tibble_3.0.4                ggplot2_3.3.2              
#[34] tidyverse_1.3.0             data.table_1.13.2          
#
#loaded via a namespace (and not attached):
#  [1] colorspace_1.4-1       hwriter_1.3.2          ellipsis_0.3.1         htmlTable_2.1.0       
#  [5] base64enc_0.1-3        fs_1.5.0               rstudioapi_0.11        bit64_4.0.5           
#  [9] AnnotationDbi_1.52.0   fansi_0.4.1            lubridate_1.7.9        xml2_1.3.2            
# [13] R.methodsS3_1.8.1      codetools_0.2-18       splines_4.0.3          doParallel_1.0.16     
# [17] impute_1.64.0          knitr_1.30             Formula_1.2-4          jsonlite_1.7.1        
# [21] broom_0.7.2            cluster_2.1.0          GO.db_3.12.1           dbplyr_2.0.0          
# [25] R.oo_1.24.0            png_0.1-7              compiler_4.0.3         httr_1.4.2            
# [29] backports_1.1.10       assertthat_0.2.1       Matrix_1.2-18          cli_2.1.0             
# [33] prettyunits_1.1.1      htmltools_0.5.0        tools_4.0.3            gtable_0.3.0          
# [37] glue_1.4.2             GenomeInfoDbData_1.2.4 rappdirs_0.3.1         Rcpp_1.0.5            
# [41] cellranger_1.1.0       vctrs_0.3.4            preprocessCore_1.52.0  rtracklayer_1.49.5    
# [45] iterators_1.0.13       xfun_0.19              rvest_0.3.6            lifecycle_0.2.0       
# [49] statmod_1.4.35         XML_3.99-0.5           MASS_7.3-53            zlibbioc_1.36.0       
# [53] scales_1.1.1           aroma.light_3.20.0     hms_0.5.3              curl_4.3              
# [57] memoise_1.1.0          gridExtra_2.3          biomaRt_2.46.0         rpart_4.1-15          
# [61] latticeExtra_0.6-29    stringi_1.5.3          RSQLite_2.2.1          foreach_1.5.1         
# [65] checkmate_2.0.0        GenomicFeatures_1.42.0 rlang_0.4.8            pkgconfig_2.0.3       
# [69] bitops_1.0-6           lattice_0.20-41        htmlwidgets_1.5.2      bit_4.0.4             
# [73] tidyselect_1.1.0       plyr_1.8.6             magrittr_1.5           R6_2.5.0              
# [77] generics_0.1.0         Hmisc_4.4-1            DelayedArray_0.16.0    DBI_1.1.0             
# [81] pillar_1.4.6           haven_2.3.1            foreign_0.8-80         withr_2.3.0           
# [85] survival_3.2-7         RCurl_1.98-1.2         nnet_7.3-14            modelr_0.1.8          
# [89] crayon_1.3.4           BiocFileCache_1.14.0   progress_1.2.2         jpeg_0.1-8.1          
# [93] locfit_1.5-9.4         grid_4.0.3             readxl_1.3.1           blob_1.2.1            
# [97] reprex_0.3.0           digest_0.6.27          R.utils_2.10.1         openssl_1.4.3         
#[101] munsell_0.5.0          askpass_1.1 